### ISSUE TITLE
torch.argmax fails for device='mps:0'

### ISSUE THREAD

**Issue by jeyendranbalakrishnan** [pytorch/pytorch#109067]
### 🐛 Describe the bug

torch.argmax gives wrong results for mps device. 

```
>>> import torch
>>> torch.backends.mps.is_available()
True
>>> device='mps:0'
>>> y = torch.tensor([14, 13, 11], dtype=torch.int, device=device)
>>> y
tensor([14, 13, 11], device='mps:0', dtype=torch.int32)
>>> i = torch.argmax(y, dim=0)
>>> i
tensor(-9223372036854775808, device='mps:0')
>>> 
```
Should give `i=0`
It works when `device='cpu'`:
```
>>> y = torch.tensor([14, 13, 11], dtype=torch.int, device='cpu')
>>> y
tensor([14, 13, 11], dtype=torch.int32)
>>> i = torch.argmax(y, dim=0)
>>> i
tensor(0)
```

### Versions

>: python collect_env.py
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.5.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.4 (main, Jul 15 2023, 12:21:29) [Clang 14.0.3 (clang-1403.0.22.14.1)] (64-bit runtime)
Python platform: macOS-13.5.1-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz

Versions of relevant libraries:
[pip3] numpy==1.25.1
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[conda] Could not collect

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by bjhargrave** [comment]
> Should give `i=0`

It does result in 0 on my M1 mac.

However, you seem to be using an intel mac.

> OS: macOS 13.5.1 (x86_64)
> Python platform: macOS-13.5.1-x86_64-i386-64bit
> CPU:
> Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz

**Comment by mikebrow** [comment]
It also works on my old iMac I had laying around.. running the latest Monterey 12.6.9 (21G726) with (Retina 5K, 27-inch, Late 2015) 3.3 GHz Quad-Core Intel Core i5, and an AMD Radeon R9 M395 w/ 2gig ram

```
>>> import torch
>>> torch.backends.mps.is_available()
True
>>> device='mps:0'
>>> y = torch.tensor([14, 13, 11], dtype=torch.int, device=device)
>>> y
tensor([14, 13, 11], device='mps:0', dtype=torch.int32)
>>> torch.argmax(y, dim=0)
tensor(0, device='mps:0')
>>> i = torch.argmax(y, dim=0)
>>> i
tensor(0, device='mps:0')

```
`
Michaels-iMac:python michaelbrown$ python3 check_env.py 
Collecting environment information...
PyTorch version: 2.2.0.dev20230920
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.9 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.6 (default, Oct 18 2022, 12:41:40)  [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-12.6.9-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i5-6600 CPU @ 3.30GHz

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.2.0.dev20230920
[pip3] torchaudio==2.2.0.dev20230920
[pip3] torchvision==0.17.0.dev20230920
[conda] Could not collect

`

**Comment by malfet** [comment]
Please try updating to pytorch-2.1.0 release candidate or latest nightly as I believe this has been fixed by https://github.com/pytorch/pytorch/pull/104374

**Comment by billvsme** [comment]
> Please try updating to pytorch-2.1.0 release candidate or latest nightly as I believe this has been fixed by #104374
@malfet 
I saw that this PR was closed, it seems like it wasn't merged.. I am using pytorch-2.1.0, but similar issues still occur.

```
>>> a = tensor([[0,0,1], [0,1,0], [0,0,1]]).type(torch.int)
>>> a = a.to("mps")
>>> a.argmax(axis=1)
tensor([0, 0, 0], device='mps:0')
>>>
```

**Comment by bjhargrave** [comment]
> >>> a.argmax(axis=1)

I don't think `axis` is an argument to the function. Do you mean `dim=1`?

**Comment by malfet** [comment]
@billvsme if you still see the problem please open a new PR, but your example seems to be working even with 2.0.1, as one can see by running the following:
```
python -c "import torch;a=torch.tensor([[0, 0, 1], [0, 1, 0], [0, 0, 1]], dtype=torch.int);print(torch.__version__, a.argmax(axis=1), a.to('mps').argmax(axis=1))"
2.0.1 tensor([2, 1, 2]) tensor([2, 1, 2], device='mps:0')
```

**Comment by billvsme** [comment]
@malfet  Yes, I am using an x86 Mac with AMD gpu. 

Here is my output:
`2.1.0 tensor([2, 1, 2]) tensor([0, 0, 0], device='mps:0')`

I reinstalled version 2.0.1, and the result is still the same
`2.0.1 tensor([2, 1, 2]) tensor([0, 0, 0], device='mps:0')`


Guessing it is due to certain differences between M1 and X86 mac or AMD gpu.

**Comment by malfet** [comment]
@billvsme do you mind opening a new issue specifically about x86 Mac (though not much we can do here other than throw an error and say that it does not work...)

**Comment by billvsme** [comment]
new issue with x86 Mac
#112186

**Comment by jhavukainen** [comment]
Closing the old issue since there's a specific one to track this for x86. There is no problem on ARM macs here.
